criu/cr-dump.c

```diff
@@ -1003,6 +1034,229 @@ static int dump_task_signals(pid_t pid, struct pstree_item *item)
 	return 0;
 }
 
+static int read_rseq_cs(pid_t tid, struct __ptrace_rseq_configuration *rseqc, struct rseq_cs *rseq_cs,
+			struct rseq *rseq)
+{
+	int ret;
+
+	/* rseq is not registered */
+	if (!rseqc->rseq_abi_pointer)
+		return 0;
+
+	/*
+	 * We need to cover the case when victim process was inside rseq critical section
+	 * at the moment when CRIU comes and seized it. We need to determine the borders
+	 * of rseq critical section at first. To achieve that we need to access thread
+	 * memory and read pointer to struct rseq_cs.
+	 *
+	 * We have two ways to access thread memory: from the parasite and using ptrace().
+	 * But it this case we can't use parasite, because if victim process returns to the
+	 * execution, on the kernel side __rseq_handle_notify_resume hook will be called,
+	 * then rseq_ip_fixup() -> clear_rseq_cs() and user space memory with struct rseq
+	 * will be cleared. So, let's use ptrace(PTRACE_PEEKDATA).
+	 */
+	ret = ptrace_peek_area(tid, rseq, decode_pointer(rseqc->rseq_abi_pointer), sizeof(struct rseq));
+	if (ret) {
+		pr_err("ptrace_peek_area(%d, %lx, %lx, %lx): fail to read rseq struct\n", tid, (unsigned long)rseq,
+		       (unsigned long)(rseqc->rseq_abi_pointer), (unsigned long)sizeof(uint64_t));
+		return -1;
+	}
+
+	if (!rseq->rseq_cs.ptr64)
```

Comment:
```
criu/cr-dump.c:1065:27: error: request for member 'ptr64' in something not a structure or union
 1065 |         if (!rseq->rseq_cs.ptr64)
      |                           ^
```

Reply:
On which Linux distribution is this reproducible? It looks like a problem with `rseq.h` system header.

Reply:
The `rseq_cs` member is no longer a union as of commit bfdf4e6208051ed7165b2e92035b4bf11f43eb63. It's a source-incompatible UAPI change.

Reply:
@fweimer-rh ah, thanks a lot!

Reply:
I've pushed a new version with fixes for that problem.

Reply:
Thanks for fixing it. I can confirm that it works now.